README.md
# base85

Base85 encoder/decoder written in native Dart.

Where base64 adds approximately 1/3, base85 only adds about 1/4. Of course there's a tradeoff. The Base85 alphabet
includes characters that might not be as friendly as the base64 alphabet. While it's still only printable characters,
the Ascii85 specification contains quotes (' and ") which needs escaping in many programming languages, and the ZeroMQ
specification contains < and > which need escaping in most (all?) SGML languages.

IPv6 encoding should only be used for encoding IPv6 addresses. When using IPv6, input for encoding must always be 16
bytes, and input for decoding must always be 20 bytes.

ZeroMQ's version (z85) require according to the specification) string input to be divisible by 5, and binary input to be
divisible by 4.

Supported encoding specifications

* [Ascii85](http://en.wikipedia.org/wiki/Ascii85)
* [ZeroMQ](http://rfc.zeromq.org/spec:32)
* [IPv6](https://tools.ietf.org/html/rfc1924)

## Example

Base85

``` dart
import 'dart:io';
import 'dart:typed_data';

import 'package:base85/base85.dart';

void main() {
  var codec = Base85Codec(Alphabets.z85);
  var encode = codec.encode(Uint8List.fromList('Hello, world!!!!'.codeUnits));
  print(encode);

  /// =>   nm=QNz.92Pz/PV8aT50L
  print(String.fromCharCodes(codec.decode(encode)));

  /// =>   Hello, world!!!!

  codec = Base85Codec(Alphabets.ascii85, AlgoType.ascii85);
  print(codec.encode(Uint8List.fromList('Hello, world'.codeUnits)));

  /// =>    <~87cURD_*#TDfTZ)~>
  print(String.fromCharCodes(codec.decode('<~87cURD_*#TDfTZ)~>')));

  /// =>    Hello, world

  codec = Base85Codec(Alphabets.IPv6, AlgoType.IPv6);
  var address = InternetAddress(
    '1080::8:800:200c:417a',
    type: InternetAddressType.IPv6,
  );
  print(codec.encode(address.rawAddress));

  /// =>    4)+k&C#VzJ4br>0wv%Yp
  ///
  var decode = codec.decode('4)+k&C#VzJ4br>0wv%Yp');
  address = InternetAddress.fromRawAddress(
    decode,
    type: InternetAddressType.IPv6,
  );
  print(address.address);

  /// =>    1080::8:800:200c:417a
}

```

CHANGELOG.md
## 1.0.4

- Fix ascii85 encoding

## 1.0.3

- Support null safety

## 1.0.2

- support [RFC 1924](https://tools.ietf.org/html/rfc1924)


## 1.0.1

- modify README


## 1.0.0

- Initial version, created by litao


test/base85_test.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:base85/base85.dart';
import 'package:test/test.dart';

void main() {
  group('z85', () {
//    var z85 = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRST' //
//        'UVWXYZ.-:+=^!/*?&<>()[]{}@%\$#';
    var codec = Base85Codec(Alphabets.z85);
    test('z85 encode', () {
      expect(
        'nm=QNz.92Pz/PV8',
        codec.encode(Uint8List.fromList('Hello, world'.codeUnits)),
      );
      expect(
        'o<}]Z',
        codec.encode(Uint8List.fromList('Man '.codeUnits)),
      );
      expect(
        'o<}]ZvpA.S',
        codec.encode(Uint8List.fromList('Man abcd'.codeUnits)),
      );
      expect(
        '',
        codec.encode(Uint8List.fromList(''.codeUnits)),
      );
      expect(
        'o<}]Zx(+zcx(!xgzFa9aB7/b}efF?GBrCHty<vdjC{3^mB0bHmvrlv8efFzABrC4raA' //
        'RphB0bKrzFa9dvr9GfvrlH7z/cXfA=k!qz//V7AV!!dx(do{B1wCTxLy%&azC)tvixx'
        'eB95Kyw/#hewGU&7zE+pvBzb98ayYQsvixJ2A=U/nwPzi%v}u^3w/\$R}y?WJ}BrCpna'
        'ARpday/tcBzkSnwN(](zE:(7zE^r<vrui@vpB4:azkn6wPzj3x(v(iz!pbczF%-nwN]'
        'B+efFIGv}xjZB0bNrwGV5cz/P}xC4Ct#zdNP{wGU]6ayPekay!&2zEEu7Abo8]B9hIm',
        codec.encode(
          Uint8List.fromList(
            ('Man is distinguished, not only by his reason, but by this singular passion ' //
                    'from other animals, which is a lust of the mind, that by a perseverance of '
                    'delight in the continued and indefatigable generation of knowledge, exceeds '
                    'the short vehemence of any carnal pleasure')
                .codeUnits,
          ),
        ),
      );
      expect(
        '00000',
        codec.encode(Uint8List.fromList([0x0, 0x0, 0x0, 0x0])),
      );
      expect(
        'HelloWorld',
        codec.encode(Uint8List.fromList(
          [0x86, 0x4F, 0xD2, 0x6F, 0xB5, 0x59, 0xF7, 0x5B],
        )),
      );
      expect(
        'JTKVSB%%)wK0E.X)V>+}o?pNmC{O&4W4b!Ni{Lh6',
        codec.encode(Uint8List.fromList([
          0x8E, 0x0B, 0xDD, 0x69, 0x76, 0x28, 0xB9, 0x1D, 0x8F, 0x24, 0x55, //
          0x87, 0xEE, 0x95, 0xC5, 0xB0, 0x4D, 0x48, 0x96, 0x3F, 0x79, 0x25,
          0x98, 0x77, 0xB4, 0x9C, 0xD9, 0x06, 0x3A, 0xEA, 0xD3, 0xB7
        ])),
      );
    });

    test('z85 decode', () {
      expect(
        'Hello, world!!!!',
        String.fromCharCodes(codec.decode('nm=QNz.92Pz/PV8aT50L')),
      );
      expect(
        'Man ',
        String.fromCharCodes(codec.decode('o<}]Z')),
      );
      expect(
        'Man abcd',
        String.fromCharCodes(codec.decode('o<}]ZvpA.S')),
      );
      expect(
        '',
        String.fromCharCodes(codec.decode('')),
      );
    });
  });

  group('ascii85', () {
//    var ascii85 = '!"#\$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVW' //
//        'XYZ[\\]^_`abcdefghijklmnopqrstu';
    var codec = Base85Codec(Alphabets.ascii85, AlgoType.ascii85);
    test('ascii85 encode', () {
      expect(
        '<~87cURD_*#TDfTZ)~>',
        codec.encode(Uint8List.fromList('Hello, world'.codeUnits)),
      );
      expect(
        '<~9jqo^~>',
        codec.encode(Uint8List.fromList('Man '.codeUnits)),
      );
      expect(
        '<~9jqo^@/~>',
        codec.encode(Uint8List.fromList('Man a'.codeUnits)),
      );
      expect(
        '<~9jqo^@:B~>',
        codec.encode(Uint8List.fromList('Man ab'.codeUnits)),
      );
      expect(
        '<~9jqo^@:E^~>',
        codec.encode(Uint8List.fromList('Man abc'.codeUnits)),
      );
      expect(
        '<~9jqo^@:E_W~>',
        codec.encode(Uint8List.fromList('Man abcd'.codeUnits)),
      );
      expect(
        '<~~>',
        codec.encode(Uint8List.fromList(''.codeUnits)),
      );
      expect(
        '<~9jqo^BlbD-BleB1DJ+*+F(f,q/0JhKF<GL>Cj@.4Gp\$d7F!,L7@<6@)/0JDEF' //
        '<G%<+EV:2F!,O<DJ+*.@<*K0@<6L(Df-\\0Ec5e;DffZ(EZee.Bl.9pF"AGXBPC'
        'si+DGm>@3BB/F*&OCAfu2/AKYi(DIb:@FD,*)+C]U=@3BN#EcYf8ATD3s@q?d\$A'
        'ftVqCh[NqF<G:8+EV:.+Cf>-FD5W8ARlolDIal(DId<j@<?3r@:F%a+D58\'ATD'
        '4\$Bl@l3De:,-DJs`8ARoFb/0JMK@qB4^F!,R<AKZ&-DfTqBG%G>uD.RTpAKYo\''
        '+CT/5+Cei#DII?(E,9)oF*2M7~>',
        codec.encode(
          Uint8List.fromList(
            ('Man is distinguished, not only by his reason, but by this singular passion ' //
                    'from other animals, which is a lust of the mind, that by a perseverance of '
                    'delight in the continued and indefatigable generation of knowledge, exceeds '
                    'the short vehemence of any carnal pleasure')
                .codeUnits,
          ),
        ),
      );
    });

    test('ascii85 decode', () {
      expect(
        'Hello, world',
        String.fromCharCodes(codec.decode('<~87cURD_*#TDfTZ)~>')),
      );
      expect(
        'Man ',
        String.fromCharCodes(codec.decode('<~9jqo^~>')),
      );
      expect(
        'Man a',
        String.fromCharCodes(codec.decode('<~9jqo^@/~>')),
      );
      expect(
        'Man ab',
        String.fromCharCodes(codec.decode('<~9jqo^@:B~>')),
      );
      expect(
        'Man abc',
        String.fromCharCodes(codec.decode('<~9jqo^@:E^~>')),
      );
      expect(
        'Man abcd',
        String.fromCharCodes(codec.decode('<~9jqo^@:E_W~>')),
      );
      expect(
        '',
        String.fromCharCodes(codec.decode('<~~>')),
      );
    });
  });

  group('IPv6', () {
    var codec = Base85Codec(Alphabets.IPv6, AlgoType.IPv6);
    test('IPv6 encode', () {
      var address = InternetAddress('1080::8:800:200c:417a');
      expect(
        '4)+k&C#VzJ4br>0wv%Yp',
        codec.encode(address.rawAddress),
      );
    });

    test('IPv6 decode', () {
      var address = InternetAddress.fromRawAddress(
        codec.decode('4)+k&C#VzJ4br>0wv%Yp'),
        type: InternetAddressType.IPv6,
      );
      expect(
        '1080::8:800:200c:417a',
        address.address,
      );
    });
  });
}


lib/src/ipv6_decoder.dart
part of base85;

var _byteMask = BigInt.from(0xff);

class Ipv6Decoder extends Converter<String, Uint8List> {
  String alphabet;
  late Uint8List _baseMap;

  Ipv6Decoder(this.alphabet) {
    _baseMap = Uint8List(256);
    _baseMap.fillRange(0, _baseMap.length, 255);
    for (var i = 0; i < alphabet.length; i++) {
      var xc = alphabet.codeUnitAt(i);
      if (_baseMap[xc] != 255) {
        throw FormatException('${alphabet[i]} is ambiguous');
      }
      _baseMap[xc] = i;
    }
  }

  @override

  /// [input] length of data must be exactly 20 bytes.
  Uint8List convert(String input) {
    if (20 != input.length) {
      throw FormatException('An encoded IPv6 is always (5/4) * 16 = 20 bytes');
    }

    var count = 0;

    var map = input.split('');
    var binary = BigInt.zero;
    for (var i = map.length - 1; i >= 0; i--) {
      var num = BigInt.from(_baseMap[map[i].codeUnits[0]]);
      var fact = BigInt.from(85).pow(count++);
      var contrib = num * fact;
      binary += contrib;
    }

    var size = (binary.bitLength + 7) >> 3;
    var bytes = Uint8List(size);
    for (var i = 0; i < size; i++) {
      bytes[size - i - 1] = (binary & _byteMask).toInt();
      binary >>= 8;
    }
    return bytes;
  }
}


lib/src/base_85_decoder.dart
part of base85;

/// Characters to allow (and ignore) in an encoded buffer
const List<int> _IGNORE_CHARS = <int>[0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x20];

/// num max value
var NUM_MAX_VALUE = pow(2, 32) - 1;

/// quad 85
var QUAD85 = pow(85, 4);

/// trio 85
var TRIO85 = pow(85, 3);

/// duo 85
var DUO85 = pow(85, 2);

/// sign 85
var SING85 = 85;

class Base85Decoder extends Converter<String, Uint8List> {
  String alphabet;
  AlgoType algo;
  late Uint8List _baseMap;

  Base85Decoder(this.alphabet, this.algo) {
    _baseMap = Uint8List(256);
    _baseMap.fillRange(0, _baseMap.length, 255);
    for (var i = 0; i < alphabet.length; i++) {
      var xc = alphabet.codeUnitAt(i);
      if (_baseMap[xc] != 255) {
        throw FormatException('${alphabet[i]} is ambiguous');
      }
      _baseMap[xc] = i;
    }
  }

  @override

  /// Decodes the specified data. If encoding is ascii85, the data is
  /// expected to start with <~ and and end with ~>.
  /// No checks are actually made for this,
  /// but output will be unexpected if this is not the case.
  ///
  /// The [input] to decode. May be a String.
  /// If ascii85, it is expected to be enclosed in <~ and ~>.
  Uint8List convert(String input) {
    if (input.isEmpty) {
      return Uint8List(0);
    }
    var bytes = Uint8List.fromList(input.codeUnits);
    var dataLength = bytes.length;
    if (algo == AlgoType.ascii85) {
      dataLength -= (ASCII85_ENC_START.length + ASCII85_ENC_END.length);
    }

    if (algo == AlgoType.z85 && dataLength % 5 != 0) {
      throw FormatException('Wrong length');
    }

    var padding = (dataLength % 5 == 0) ? 0 : 5 - dataLength % 5;

    var bufferStart = (algo == AlgoType.ascii85) ? ASCII85_ENC_START.length : 0;
    var bufferEnd = bufferStart + dataLength;

    var result = Uint8List(4 * ((bufferEnd - bufferStart) / 5).ceil());

    var nextValidByte = (index) {
      if (index < bufferEnd) {
        while (_IGNORE_CHARS.contains(bytes[index])) {
          padding = (padding + 1) % 5;
          index++; // skip newline character
        }
      }
      return index;
    };

    var writeIndex = 0;
    for (var i = bufferStart; i < bufferEnd;) {
      var num = 0;
      var starti = i;

      i = nextValidByte(i);
      num = (_baseMap[bytes[i]]) * (QUAD85 as int);

      i = nextValidByte(i + 1);
      num += (i >= bufferEnd ? 84 : _baseMap[bytes[i]]) * (TRIO85 as int);

      i = nextValidByte(i + 1);
      num += (i >= bufferEnd ? 84 : _baseMap[bytes[i]]) * (DUO85 as int);

      i = nextValidByte(i + 1);
      num += (i >= bufferEnd ? 84 : _baseMap[bytes[i]]) * SING85;

      i = nextValidByte(i + 1);
      num += (i >= bufferEnd ? 84 : _baseMap[bytes[i]]);

      i = nextValidByte(i + 1);

      if (algo == AlgoType.z85 && starti + 5 != i) {
        throw FormatException('Wrong length');
      }

      if (num > NUM_MAX_VALUE || num < 0) {
        throw FormatException('Bogus data');
      }
      result[writeIndex] = (num >> 24);
      result[writeIndex + 1] = (num >> 16);
      result[writeIndex + 2] = (num >> 8);
      result[writeIndex + 3] = (num & 0xff);
      writeIndex += 4;
    }

    return result.sublist(0, writeIndex - padding);
  }
}


lib/src/base_85_encoder.dart
part of base85;

var ASCII85_ENC_START = '<~';
var ASCII85_ENC_END = '~>';

class Base85Encoder extends Converter<Uint8List, String> {
  final String alphabet;
  final AlgoType algo;

  Base85Encoder(this.alphabet, this.algo);

  @override

  /// Encodes the specified data. If algo is [AlgoType.ascii85],
  /// the encoded data will be prepended with <~ and appended with ~>.
  /// The [bytes] to encode, may be a [Uint8List]
  /// return A String with the encoded [bytes].
  String convert(Uint8List bytes) {
    if (algo == AlgoType.z85 && bytes.length % 4 != 0) {
      throw FormatException('Wrong length');
    }
    var padding = (bytes.length % 4 == 0) ? 0 : 4 - bytes.length % 4;

    var result = '';
    for (var i = 0; i < bytes.length; i += 4) {
      /// 32-bit number of the current 4 bytes (padded with 0 as necessary)
      var num = ((bytes[i] << 24).toUnsigned(32)) +
          (((i + 1 >= bytes.length ? 0 : bytes[i + 1]) << 16).toUnsigned(32)) +
          (((i + 2 >= bytes.length ? 0 : bytes[i + 2]) << 8).toUnsigned(32)) +
          (((i + 3 >= bytes.length ? 0 : bytes[i + 3]) << 0).toUnsigned(32));

      /// Create 5 characters from '!' to 'u' alphabet
      var block = <String>[];
      for (var j = 0; j < 5; ++j) {
        block.insert(0, alphabet[num % 85]);
        num = num ~/ 85;
      }

      var r = block.join('');
      if (r == '!!!!!' && algo == AlgoType.ascii85) {
        r = 'z';
      }

      /// And append them to the result
      result += r;
    }

    return ((algo == AlgoType.ascii85) ? ASCII85_ENC_START : '') +
        result.substring(0, result.length - padding) +
        ((algo == AlgoType.ascii85) ? ASCII85_ENC_END : '');
  }
}


example/base85_example.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:base85/base85.dart';

void main() {
  var codec = Base85Codec(Alphabets.z85);
  String encode = codec.encode(Uint8List.fromList('Hello, world!!!!'.codeUnits));
  print(encode);

  /// =>   nm=QNz.92Pz/PV8aT50L
  print(String.fromCharCodes(codec.decode(encode)));

  /// =>   Hello, world!!!!

  codec = Base85Codec(Alphabets.ascii85, AlgoType.ascii85);
  print(codec.encode(Uint8List.fromList('Hello, world'.codeUnits)));

  /// =>    <~87cURD_*#TDfTZ)~>
  print(String.fromCharCodes(codec.decode('<~87cURD_*#TDfTZ)~>')));

  /// =>    Hello, world

  codec = Base85Codec(Alphabets.IPv6, AlgoType.IPv6);
  var address = InternetAddress(
    '1080::8:800:200c:417a',
    type: InternetAddressType.IPv6,
  );
  print(codec.encode(address.rawAddress));

  /// =>    4)+k&C#VzJ4br>0wv%Yp
  ///
  Uint8List decode = codec.decode('4)+k&C#VzJ4br>0wv%Yp');
  address = InternetAddress.fromRawAddress(
    decode,
    type: InternetAddressType.IPv6,
  );
  print(address.address);

  /// =>    1080::8:800:200c:417a
}
